src/main/src/main/java/org/geoserver/catalog/ResourcePool.java

```diff
@@ -1382,15 +1389,22 @@ else if (ppolicy == ProjectionPolicy.NONE && nativeCRS != null) {
             }
 
             // return a normal
```

Comment:
This comment needs to be updated, a comments explaining a bit was going on would be welcome :smile_cat:.

Reply:
Agreed, and would add a note about the order of operations, why first create the geoserver feature source and then run the wrappers after it, instead of the opposite. I believe it has to be this way to match the feature type building process?
However picture this case, a wrapper is building a geometry type that was not there. GeoServer feature source applies the projection handling, so doing it before, will make the code lose it. _Maybe_ it's better to apply the callbacks before any sort of CRS handling is done, both here and computing the feature type, since the CRS handling is one of the user visible/user configurable things. Let's talk about it at least?

Reply:
Good catch @aaime (and thank you for the offline explanation :wink:). So @imranrajjad, like me you may have not noticed that this code:
```java
            // return a normal
            return GeoServerFeatureLocking.create(
                    fs,
                    new GeoServerFeatureSource.Settings(
                            schema,
                            info.filter(),
                            resultCRS,
                            info.getProjectionPolicy().getCode(),
                            getTolerance(info),
                            info.getMetadata()));
```

Is actually wrapping the original feature source ``fs`` with a wrapper that will provide extra GeoServer functionalities, for example, the ability to define a layer with a CQL filter or the ability of forcing a layer for publishing different from the native one. 

Hence this extension wrapping  should happen before this is done, and we shoudl actually wrap the original feature source, the variable ``fs``.

The integration tests should also be extended to tests this situations:
- A layer whit a CQL filter associated, querying the layer using WMS or WFs should filter out the features not matching the layer associated CQL
- A layer with the declared CRS different form the native one, querying the layer iwht either WFS or WMS should return featrues expressed in the declared SRS not in the native one.

Reply:
@nmco its done